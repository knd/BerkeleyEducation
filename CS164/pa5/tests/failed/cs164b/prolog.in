def range(start, stop) {
    def i = start
    lambda() {
        if(i < stop) {
            def cur = i
            i = i + 1
            cur
        } else {
            null
        }
    }
}

def take_from(list, n) {
    def res = {}
    def i = 0
    for (x in list) {
        if (i >= n) {
            append(res, x)
        }
        i = i + 1
    }
    res
}

def copy(list) {
    def res = {}
    for (x in list) {
        append(res, x)
    }
    res
}

def wrap(lam, arg) {
    def co = coroutine(lam)
    lambda() { resume(co, arg) }
}

def isVar(t)  {
    t.typ == 'var'    
}

def isCst(t)  {
    t.typ == 'cst'    
}

def isClause(t)  {
    t.typ == 'cls'    
}


def contains(term, var) {
    def contains0(clause, var) {
        def res = False
        for (term in clause) {
            res = res || contains(term, var)
        }
        res
    }
    
    if (isCst(term)) {
        False
    } else { if (isVar(term)) {
        term.name == var
    } else { if (isClause(term)) {
        contains0(term.ref, var)
    } else {
        error "contains: Unknown term type: " + term.typ 
    }}}
}

def subst_term(term, slist) {
    def newterm = {typ = term.typ}
    if (isVar(term)) {
        newterm.name = term.name
        for (s in slist) {
            if (s.from == term.name) {
                newterm = s.to
            }
        }
    } else { if (isClause(term)) {
        newterm.ref = subst_clause(term.ref, slist)
    } else { if (isCst(term)) {
        newterm.name = term.name
    } else {
        error "subst0: Unknown term type: " + term.typ
    }}}
    newterm
}

def subst_clause(clause, slist) {
    def newclause = {name = clause.name, negated = clause.negated}
    for (i in range(0, len(clause))) {
        newclause[i] = subst_term(clause[i], slist)
    }
    newclause
}

def subst(clauses, slist) {
    def newclist = {}
    for (c in clauses) {
        def newc = subst_clause(c, slist)
        append(newclist, newc)
    }
    newclist
}

def merge(mgu1, mgu2) {
    if (mgu1 == null || mgu2 == null) {
        null
    } else {
        def res = mgu1
        for (x in mgu2) {
            append(res, x)
        }
        res
    }
}

def unify(clause1, clause2) {
    if (clause1.name != clause2.name || len(clause1) != len(clause2)) {
        null
    } else {
        def res = {}
        def i = 0
        while(i < len(clause1)) {
            def term1 = subst_term(clause1[i], res)
            def term2 = subst_term(clause2[i], res)
            #print "Unifying " + term1 + " and " + term2
            if (isCst(term1) && isCst(term2) && term1.name == term2.name) {
                # Nothing to do here
            } else { if (isVar(term1) && (not(contains(term2, term1.name)) || isVar(term2))) {
                def s = {from = term1.name , to = subst_term(term2, res)}
                append(res, s)
            } else { if (isVar(term2) && (not(contains(term1, term2.name)) || isVar(term1))) {
                def s = {from = term2.name, to = subst_term(term1, res)}
                append(res, s)
            } else { if (isClause(term1) && isClause(term2)) {
                def innermgu = unify(term1.ref, term2.ref)
                if (innermgu == null) {
                    i = len(clause1)
                    res = null
                } else {
                    merge(res, innermgu)
                }
            } else {
                # No unifier
                i = len(clause1)
                res = null
            }}}}
            i = i + 1
        }
        res
    }
}

def process(goal) {

    # Returns null if a solution to the goal does indeed exist. If a solution
    # doesn't exist, determine what value(s) this function may return while
    # maintaining your program's overall correctness.
    #
    # goal - the goal for which to check whether a solution doesn't exist
    #
    def negation(goal) {
        def newgoal = {name = goal.name, negated = False}
        for (term in goal) {
            append(newgoal, term)
        }
        if (resume(coroutine(process), newgoal) == null) {
            yield({})
        }
        null

        # [~7 lines]
        # The documentation for this function explains the desired behavior.
    }

    # Returns an MGU that simultaneously satisfies all of the given clauses.
    # If no solution exists, this function returns null.
    #
    # clauses - a list of clauses to satisfy. Variables in these clauses have
    #           already been substituted with those in `mgus', where possible.
    #    mgus - an MGU, possibly built up from several MGUs, that have been
    #           applied to the clauses and is a subset of the MGU to return if
    #           a solution exists
    #   depth - the index of the clause to process
    #
    def conjunction(clauses, mgus, depth) {
        def goal = clauses[depth]
        def solutions = wrap(process, goal)
        for (mgu in solutions) {
            if (depth == len(clauses) - 1) {
                yield(merge(mgu, mgus))
            } else {
                conjunction(subst(clauses, mgu), merge(mgu, mgus), depth + 1)
            }
        }
        null

        # [~10 lines]
        # The current goal is the clause that we want to satisfy.

        # Each MGU for the goal may be part of a larger solution, which also
        # includes any other MGUs generated for this conjunction.

        # If we have processed all of the clauses then we know a set of MGUs
        # that simultaneously satisfy all of the clauses.

        # Otherwise, we still need to ensure that the remaining clauses are
        # satisfiable, given that we apply the recently generated MGU and
        # include it as part of our larger potential solution.
    }

    #print "Goal: " + goal
    if (goal.negated) {
        negation(goal)
    } else {
        for (rule in prog) {
            def mgu = unify(rule.head, goal)
            if (mgu != null) {
                if (rule.body == null) {
                    yield(mgu)
                } else {
                    def partialMgus = wrap(lambda(_) {
                      conjunction(subst(rule.body, mgu), {}, 0)
                    }, null)
                    for (m in partialMgus) {
                      yield(merge(m, mgus))
                    }
                }
            }
        }
        null
    }

    # [~15 lines]
    # If the goal is negated, we want to see if no solution exists.

    # Otherwise, we need to apply disjunction; is at least one rule (including
    # axioms) satisfied?

    # The list `prog' contains all of the rules and axioms of the program.

    # As shown in lecture, we want to see if each rule's head can unify with
    # our current goal.

    # Unifications with axioms (body-less rules) provide a base case in that
    # the MGU is a complete solution for the current goal.

    # Otherwise, if our rule has a body, we must apply conjunction. This
    # entails applying the MGU (a partial solution) to the clauses in the rule
    # body and then seeing if a simultaneous solution exists for all of them. Be
    # careful here -- once we substitute as many variables in the clause with
    # those in the MGU as much as possible, the MGU in this scope is no longer
    # relevant.

    # If none of the rules' heads unify with the goal, then no solution exists.
}


# Main loop
def co = coroutine(process)
def i = 1
print "Asking for solution " + i
def s = resume(co, query)
print "Answer: " + s

while (s != null) {
    i = i + 1
    print "Asking for solution " + i
    s = resume(co, query)
    print "Answer: " + s
}



