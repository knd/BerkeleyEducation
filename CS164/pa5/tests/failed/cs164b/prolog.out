[('def', 'range', ('lambda', ['start', 'stop'], [('def', 'i', ('var', 'start')), ('exp', ('lambda', [], [('if', ('<', ('var', 'i'), ('var', 'stop')), [('def', 'cur', ('var', 'i')), ('asgn', 'i', ('+', ('var', 'i'), ('int-lit', 1))), ('exp', ('var', 'cur'))], [('exp', ('null',))])]))])), ('def', 'take_from', ('lambda', ['list', 'n'], [('def', 'res', ('dict-lit', [])), ('def', 'i', ('int-lit', 0)), ('for', 'x', ('var', 'list'), [('if', ('>=', ('var', 'i'), ('var', 'n')), [('exp', ('call', ('var', 'append'), [('var', 'res'), ('var', 'x')]))], None), ('asgn', 'i', ('+', ('var', 'i'), ('int-lit', 1)))]), ('exp', ('var', 'res'))])), ('def', 'copy', ('lambda', ['list'], [('def', 'res', ('dict-lit', [])), ('for', 'x', ('var', 'list'), [('exp', ('call', ('var', 'append'), [('var', 'res'), ('var', 'x')]))]), ('exp', ('var', 'res'))])), ('def', 'wrap', ('lambda', ['lam', 'arg'], [('def', 'co', ('coroutine', ('var', 'lam'))), ('exp', ('lambda', [], [('exp', ('resume', ('var', 'co'), ('var', 'arg')))]))])), ('def', 'isVar', ('lambda', ['t'], [('exp', ('==', ('get', ('var', 't'), ('string-lit', 'typ')), ('string-lit', 'var')))])), ('def', 'isCst', ('lambda', ['t'], [('exp', ('==', ('get', ('var', 't'), ('string-lit', 'typ')), ('string-lit', 'cst')))])), ('def', 'isClause', ('lambda', ['t'], [('exp', ('==', ('get', ('var', 't'), ('string-lit', 'typ')), ('string-lit', 'cls')))])), ('def', 'contains', ('lambda', ['term', 'var'], [('def', 'contains0', ('lambda', ['clause', 'var'], [('def', 'res', ('var', 'False')), ('for', 'term', ('var', 'clause'), [('asgn', 'res', ('||', ('var', 'res'), ('call', ('var', 'contains'), [('var', 'term'), ('var', 'var')])))]), ('exp', ('var', 'res'))])), ('if', ('call', ('var', 'isCst'), [('var', 'term')]), [('exp', ('var', 'False'))], [('if', ('call', ('var', 'isVar'), [('var', 'term')]), [('exp', ('==', ('get', ('var', 'term'), ('string-lit', 'name')), ('var', 'var')))], [('if', ('call', ('var', 'isClause'), [('var', 'term')]), [('exp', ('call', ('var', 'contains0'), [('get', ('var', 'term'), ('string-lit', 'ref')), ('var', 'var')]))], [('error', ('+', ('string-lit', 'contains: Unknown term type: '), ('get', ('var', 'term'), ('string-lit', 'typ'))))])])])])), ('def', 'subst_term', ('lambda', ['term', 'slist'], [('def', 'newterm', ('dict-lit', [('typ', ('get', ('var', 'term'), ('string-lit', 'typ')))])), ('if', ('call', ('var', 'isVar'), [('var', 'term')]), [('put', ('var', 'newterm'), ('string-lit', 'name'), ('get', ('var', 'term'), ('string-lit', 'name'))), ('for', 's', ('var', 'slist'), [('if', ('==', ('get', ('var', 's'), ('string-lit', 'from')), ('get', ('var', 'term'), ('string-lit', 'name'))), [('asgn', 'newterm', ('get', ('var', 's'), ('string-lit', 'to')))], None)])], [('if', ('call', ('var', 'isClause'), [('var', 'term')]), [('put', ('var', 'newterm'), ('string-lit', 'ref'), ('call', ('var', 'subst_clause'), [('get', ('var', 'term'), ('string-lit', 'ref')), ('var', 'slist')]))], [('if', ('call', ('var', 'isCst'), [('var', 'term')]), [('put', ('var', 'newterm'), ('string-lit', 'name'), ('get', ('var', 'term'), ('string-lit', 'name')))], [('error', ('+', ('string-lit', 'subst0: Unknown term type: '), ('get', ('var', 'term'), ('string-lit', 'typ'))))])])]), ('exp', ('var', 'newterm'))])), ('def', 'subst_clause', ('lambda', ['clause', 'slist'], [('def', 'newclause', ('dict-lit', [('name', ('get', ('var', 'clause'), ('string-lit', 'name'))), ('negated', ('get', ('var', 'clause'), ('string-lit', 'negated')))])), ('for', 'i', ('call', ('var', 'range'), [('int-lit', 0), ('len', ('var', 'clause'))]), [('put', ('var', 'newclause'), ('var', 'i'), ('call', ('var', 'subst_term'), [('get', ('var', 'clause'), ('var', 'i')), ('var', 'slist')]))]), ('exp', ('var', 'newclause'))])), ('def', 'subst', ('lambda', ['clauses', 'slist'], [('def', 'newclist', ('dict-lit', [])), ('for', 'c', ('var', 'clauses'), [('def', 'newc', ('call', ('var', 'subst_clause'), [('var', 'c'), ('var', 'slist')])), ('exp', ('call', ('var', 'append'), [('var', 'newclist'), ('var', 'newc')]))]), ('exp', ('var', 'newclist'))])), ('def', 'merge', ('lambda', ['mgu1', 'mgu2'], [('if', ('||', ('==', ('var', 'mgu1'), ('null',)), ('==', ('var', 'mgu2'), ('null',))), [('exp', ('null',))], [('def', 'res', ('var', 'mgu1')), ('for', 'x', ('var', 'mgu2'), [('exp', ('call', ('var', 'append'), [('var', 'res'), ('var', 'x')]))]), ('exp', ('var', 'res'))])])), ('def', 'unify', ('lambda', ['clause1', 'clause2'], [('if', ('||', ('!=', ('get', ('var', 'clause1'), ('string-lit', 'name')), ('get', ('var', 'clause2'), ('string-lit', 'name'))), ('!=', ('len', ('var', 'clause1')), ('len', ('var', 'clause2')))), [('exp', ('null',))], [('def', 'res', ('dict-lit', [])), ('def', 'i', ('int-lit', 0)), ('while', ('<', ('var', 'i'), ('len', ('var', 'clause1'))), [('def', 'term1', ('call', ('var', 'subst_term'), [('get', ('var', 'clause1'), ('var', 'i')), ('var', 'res')])), ('def', 'term2', ('call', ('var', 'subst_term'), [('get', ('var', 'clause2'), ('var', 'i')), ('var', 'res')])), ('if', ('&&', ('&&', ('call', ('var', 'isCst'), [('var', 'term1')]), ('call', ('var', 'isCst'), [('var', 'term2')])), ('==', ('get', ('var', 'term1'), ('string-lit', 'name')), ('get', ('var', 'term2'), ('string-lit', 'name')))), [], [('if', ('&&', ('call', ('var', 'isVar'), [('var', 'term1')]), ('||', ('call', ('var', 'not'), [('call', ('var', 'contains'), [('var', 'term2'), ('get', ('var', 'term1'), ('string-lit', 'name'))])]), ('call', ('var', 'isVar'), [('var', 'term2')]))), [('def', 's', ('dict-lit', [('from', ('get', ('var', 'term1'), ('string-lit', 'name'))), ('to', ('call', ('var', 'subst_term'), [('var', 'term2'), ('var', 'res')]))])), ('exp', ('call', ('var', 'append'), [('var', 'res'), ('var', 's')]))], [('if', ('&&', ('call', ('var', 'isVar'), [('var', 'term2')]), ('||', ('call', ('var', 'not'), [('call', ('var', 'contains'), [('var', 'term1'), ('get', ('var', 'term2'), ('string-lit', 'name'))])]), ('call', ('var', 'isVar'), [('var', 'term1')]))), [('def', 's', ('dict-lit', [('from', ('get', ('var', 'term2'), ('string-lit', 'name'))), ('to', ('call', ('var', 'subst_term'), [('var', 'term1'), ('var', 'res')]))])), ('exp', ('call', ('var', 'append'), [('var', 'res'), ('var', 's')]))], [('if', ('&&', ('call', ('var', 'isClause'), [('var', 'term1')]), ('call', ('var', 'isClause'), [('var', 'term2')])), [('def', 'innermgu', ('call', ('var', 'unify'), [('get', ('var', 'term1'), ('string-lit', 'ref')), ('get', ('var', 'term2'), ('string-lit', 'ref'))])), ('if', ('==', ('var', 'innermgu'), ('null',)), [('asgn', 'i', ('len', ('var', 'clause1'))), ('asgn', 'res', ('null',))], [('exp', ('call', ('var', 'merge'), [('var', 'res'), ('var', 'innermgu')]))])], [('asgn', 'i', ('len', ('var', 'clause1'))), ('asgn', 'res', ('null',))])])])]), ('asgn', 'i', ('+', ('var', 'i'), ('int-lit', 1)))]), ('exp', ('var', 'res'))])])), ('def', 'process', ('lambda', ['goal'], [('def', 'negation', ('lambda', ['goal'], [('def', 'newgoal', ('dict-lit', [('name', ('get', ('var', 'goal'), ('string-lit', 'name'))), ('negated', ('var', 'False'))])), ('for', 'term', ('var', 'goal'), [('exp', ('call', ('var', 'append'), [('var', 'newgoal'), ('var', 'term')]))]), ('if', ('==', ('resume', ('coroutine', ('var', 'process')), ('var', 'newgoal')), ('null',)), [('exp', ('yield', ('dict-lit', [])))], None), ('exp', ('null',))])), ('def', 'conjunction', ('lambda', ['clauses', 'mgus', 'depth'], [('def', 'goal', ('get', ('var', 'clauses'), ('var', 'depth'))), ('def', 'solutions', ('call', ('var', 'wrap'), [('var', 'process'), ('var', 'goal')])), ('for', 'mgu', ('var', 'solutions'), [('if', ('==', ('var', 'depth'), ('-', ('len', ('var', 'clauses')), ('int-lit', 1))), [('exp', ('yield', ('call', ('var', 'merge'), [('var', 'mgu'), ('var', 'mgus')])))], [('exp', ('call', ('var', 'conjunction'), [('call', ('var', 'subst'), [('var', 'clauses'), ('var', 'mgu')]), ('call', ('var', 'merge'), [('var', 'mgu'), ('var', 'mgus')]), ('+', ('var', 'depth'), ('int-lit', 1))]))])]), ('exp', ('null',))])), ('if', ('get', ('var', 'goal'), ('string-lit', 'negated')), [('exp', ('call', ('var', 'negation'), [('var', 'goal')]))], [('for', 'rule', ('var', 'prog'), [('def', 'mgu', ('call', ('var', 'unify'), [('get', ('var', 'rule'), ('string-lit', 'head')), ('var', 'goal')])), ('if', ('!=', ('var', 'mgu'), ('null',)), [('if', ('==', ('get', ('var', 'rule'), ('string-lit', 'body')), ('null',)), [('exp', ('yield', ('var', 'mgu')))], [('def', 'partialMgus', ('call', ('var', 'wrap'), [('lambda', ['_'], [('exp', ('call', ('var', 'conjunction'), [('call', ('var', 'subst'), [('get', ('var', 'rule'), ('string-lit', 'body')), ('var', 'mgu')]), ('dict-lit', []), ('int-lit', 0)]))]), ('null',)])), ('for', 'm', ('var', 'partialMgus'), [('exp', ('yield', ('call', ('var', 'merge'), [('var', 'm'), ('var', 'mgus')])))])])], None)]), ('exp', ('null',))])])), ('def', 'co', ('coroutine', ('var', 'process'))), ('def', 'i', ('int-lit', 1)), ('print', ('+', ('string-lit', 'Asking for solution '), ('var', 'i'))), ('def', 's', ('resume', ('var', 'co'), ('var', 'query'))), ('print', ('+', ('string-lit', 'Answer: '), ('var', 's'))), ('while', ('!=', ('var', 's'), ('null',)), [('asgn', 'i', ('+', ('var', 'i'), ('int-lit', 1))), ('print', ('+', ('string-lit', 'Asking for solution '), ('var', 'i'))), ('asgn', 's', ('resume', ('var', 'co'), ('var', 'query'))), ('print', ('+', ('string-lit', 'Answer: '), ('var', 's')))])]
